### PR TITLE
Add "Latest Posts" Block

### PR DESCRIPTION
**Pull request by lamosty**
@mtias suggested to me to create a simple block to get latest posts with the WP REST API. It should also serve as an example for other developers so they can start experimenting with the API and create some cool blocks.

I've been studying Gutenberg (pretty good editor!) and come up with something really simple. It's still a work in progress since I want to fix a few things and maybe make it a bit better.

I'd like to also open a discussion about some of the things here:
- naming: I prefixed it with `rest-api-` since we could have another 'latest posts' block without using rest api
- new category called `rest-api`: are we planning to have more such components using rest api? If yes, I think it would be nice to give them their own category
- getting the posts/using the rest api: I'm using the [Backbone rest api js client](https://developer.wordpress.org/rest-api/using-the-rest-api/backbone-javascript-client/) as it seems to be the officially supported one. It offers a few niceties and it's quite easy to use but: isn't Backbone a bit dated? (I've never used it). Are we okay with this or should we create some custom rest api wrapper/client?

Asking for help: after saving a post with this latest posts block and then reloading the saved post, instead of having the block parsed and displayed as other blocks, it shows with the Gutenberg tags and HTML code. Can you spot what I'm doing wrong, please? It works fine when seeing the post on the front-end though.

Am I using the `attributes` of `registerBlock` properly? I use them to store the latest posts retrieved from DB but don't define them in `registerBlock`.

Thanks!

## Nice to have/TODO
- [ ] have some UI so users can filter the posts by number of them to show, author, categories, etc

### CODE REVIEW

**Comment by youknowriad** [comment]
Thanks for doing this. 

At the moment, this block will shows the latest post at the moment where we save the block, if we create a new post after that. This new post is not included in the frontend. I think this post is a good opportunity to leverage the Server Side Rendering to address this.

**Comment by lamosty** [comment]
> I think this post is a good opportunity to leverage the Server Side Rendering to address this.

Yeah, I've seen your PR but haven't looked into it more. I also agree since it's an expected behavior to always display the latest posts, especially when using REST API (ie it's 'dynamic').

**Comment by mtias** [comment]
Thanks for giving this a try!

> I also agree since it's an expected behavior to always display the latest posts, especially when using REST API (ie it's 'dynamic').

Definitely, this is a great example of a purely dynamic block in my book.

File: blocks/library/rest-api-latest-posts/index.js [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+ * WordPress dependencies
+ */
+import Placeholder from 'components/placeholder';
+
+/**
+ * Internal dependencies
+ */
+import { registerBlock } from '../../api';
+
+function getLatestPosts( postsToShow = 5 ) {
```
Comment (mtias) [inline blocks/library/rest-api-latest-posts/index.js]: Thinking... this could be a good example for breaking some logic into a separate file. Perhaps `rest-api-latest-posts/data.js`?

File: blocks/library/rest-api-latest-posts/index.js [hunk 2]
```diff
@@ -0,0 +1,71 @@
+/**
```
Comment (mtias) [inline blocks/library/rest-api-latest-posts/index.js]: Let's call this block just `latest-posts`.

**Comment by westonruter** [comment]
There's good overlap here with the Recent Posts widget.

**Comment by jasmussen** [comment]
Sweet deal, and awesome work!

I was so inspired I took a stab at the design of this, as suggested in your ToDo. Note: the following is very mockuppy and conceptual, and I expect some pushback here. The implementation is also probably something that can happen best in a separate PR, but nonetheless to keep the discussion grouped for now, here it is. 

Neutral:

recent posts neutral

Selected:

recent posts selected

This leverages both the "block-first" UI of putting the most basic stuff right on the block, and the more advanced stuff in the Inspector on the right. For this concept, I arbitrarily decided to put the post date and number of posts in the inspector, but this was also done in order to grease the discussion on what we put there (see #672 ).

**Comment by mtias** [comment]
@jasmussen I like this, but I think we could also avoid it looking like a form if we just made the title an optional editable field like caption/quote-source with placeholder text instead of a label and an input.

**Comment by jasmussen** [comment]
> I like this, but I think we could also avoid it looking like a form if we just made the title an optional editable field like caption/quote-source with placeholder text instead of a label and an input.

This was my first instinct as well. I do think we'll want some further iteration here, as to what feels right. But I specifically like the gray background which is a callback to placeholders. Not only does this help the block feel distinct from the others, in that it's dynamic, but it also literally "grays" the content, implying you can't edit it directly.

**Comment by lamosty** [comment]
Question: can `attributes` contain an object? For example, I'm storing the retrieved posts as JS objects in `attributes` and upon calling `save`, they get rendered into the comment as:

selection_302

However, it seems like Gutenberg doesn't like if attribute is `"[object Object]"` and doesn't render  the block properly when loading a saved instance.

**Comment by youknowriad** [comment]
@lamosty `attributes` should not store objects (At least it can't for now). I'd suggest you store the posts in a local state (the `edit` function can be a React Component). And I'd just return `null` in the save function right now.

**Comment by lamosty** [comment]
In 497cee8, I added simple server-side rendering. However, I had to rename the component to `core/latestposts` from `core/latest-posts` because if there's a hyphen in the name, the server-side parser will ignore it. I think that's a bug and opened a new issue: #882.

Now, I have some more questions :):
- when doing server-side rendering of a block, we have to write the render logic again and try to match it as closely as possible to how it works in JS. Are there some plans/ideas to add ability to write just one render logic and have it render on both client and server?
- is it possible to use the same code that runs in the editor to render stuff on site's front-end? I.e. use the `edit` method to work on front-end? If not, are there any plans to support this?

**Comment by youknowriad** [comment]
> when doing server-side rendering of a block, we have to write the render logic again and try to match it as closely as possible to how it works in JS. Are there some plans/ideas to add ability to write just one render logic and have it render on both client and server?

Yes, @notnownikki proposed a SSR endpoint see (https://github.com/WordPress/gutenberg/issues/780) and implemented it in the WIP pr #816 
I think this endpoint should be extracted to its own PR

> is it possible to use the same code that runs in the editor to render stuff on site's front-end? I.e. use the edit method to work on front-end? If not, are there any plans to support this?

For now, the frontend is just a raw HTML, saved from the editor (the `save`function can share some of the `edit` elements) or Server side rendered. I guess you're proposing to use a React component as an output, It's the first time I'm seeing this proposal, it raises so many questions, maybe worth creating an issue to discuss this possibility (but I don't see this for a v1 personally)

File: blocks/library/latest-posts/index.js [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+			};
+
+			let { poststoshow } = this.props.attributes;
+
+			if ( ! poststoshow ) {
+				poststoshow = 5;
+
+				this.props.setAttributes( {
+					poststoshow
+				} );
```
Comment (youknowriad) [inline blocks/library/latest-posts/index.js]: I haved mixed feelings here, we're calling `setAttributes` in the constructor which could change the markup directly. I think it's better to use `defaultAttributes` property for this (see usage in the `table` block I think).

Comment (lamosty) [inline blocks/library/latest-posts/index.js]: Ah, I didn't know we have `defaultAttributes`. Thanks!

Comment (notnownikki) [inline blocks/library/latest-posts/index.js]: I'm happy to put the SSR endpoint into its own PR, can start on it now if it's wanted :)

**Comment by paaljoachim** [comment]
Here are the various options that are used in the Genesis featured posts widget. I am posting it to give additional information on options that might be good to include. 

genesis-featured-posts

File: blocks/library/latest-posts/index.js (hunk 3 above)
Comment (lamosty) [inline blocks/library/latest-posts/index.js]: > I'm happy to put the SSR endpoint into its own PR, can start on it now if it's wanted :)

No rush, I think I'll leave the SSR part as it is now and we can improve it in a future PR. :)

**Comment by lamosty** [comment]
> Here are the various options that are used in the Genesis featured posts widget. I am posting it to give additional information on options that might be good to include.

Wow, that's quite a list!

I think I'll try to keep this simple though and try to implement @jasmussen's design for now. We can always improve. :)

**Comment by paaljoachim** [comment]
Yeah too many options in there. That list is good to have as some of the options might be good to add at a later date.

File: blocks/library/latest-posts/data.js [hunk 4]
```diff
@@ -0,0 +1,18 @@
+/**
+ * Returns a Promise with the latest posts or an error on failure.
+ *
+ * @param   {Number} postsToShow
+ * @returns {Object}
```
Comment (westonruter) [inline blocks/library/latest-posts/data.js]: Instead of `{Object}` this can be `{wp.api.collections.Posts}`

File: blocks/library/latest-posts/index.php [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+		'numberposts' => $postsToShow,
+		'post_status' => 'publish'
+	) );
+
+	$posts_content = '';
+
+	foreach( $recent_posts as $post ) {
+		$post_permalink = get_permalink( $post['ID'] );
+
+		$posts_content .= "<li><a href='{$post_permalink}'>{$post['post_title']}</a></li>\n";
```
Comment (westonruter) [inline blocks/library/latest-posts/index.php]: Instead of `$post['post_title']` this should use `get_the_title( $post['ID'] )` so that filters apply.

File: index.php [hunk 6]
```diff
@@ -473,3 +476,8 @@ function the_gutenberg_project() {
 	</div>
 	<?php
 }
+
+function gutenberg_load_blocks_server_side_redering() {
```
Comment (westonruter) [inline index.php]: Typo `s/redering/rendering/`

File: blocks/library/latest-posts/index.php [hunk 7]
```diff
@@ -0,0 +1,37 @@
+<?php
+
+function gutenberg_block_core_latest_posts( $attributes ) {
+	$postsToShow = 5;
```
Comment (westonruter) [inline blocks/library/latest-posts/index.php]: Shouldn't this be `$posts_to_show`?

**Comment by nylen** [comment]
After #849 which added a bunch of tests for block parsing and serialization, this PR should probably be rebased against `master`, either because the build will fail upon merge, or because this PR can add some new tests using this structure ([details here](https://github.com/WordPress/gutenberg/blob/484c146da24c8dac8232452f559161d58023c5fd/blocks/test/fixtures/README.md)).

File: blocks/library/latest-posts/index.js [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+import Placeholder from 'components/placeholder';
+
+/**
+ * Internal dependencies
+ */
+import { registerBlockType } from '../../api';
+import { getLatestPosts } from './data.js';
+
+registerBlockType( 'core/latestposts', {
+	title: wp.i18n.__( 'Latest Posts' ),
```
Comment (youknowriad) [inline blocks/library/latest-posts/index.js]: Minor: Should we be consistent about importing `wp.*` and use an `import { __ } from 'i18n';` here?

File: blocks/library/latest-posts/index.js [hunk 9]
```diff
@@ -0,0 +1,75 @@
+/**
+ * WordPress dependencies
+ */
+import Placeholder from 'components/placeholder';
```
Comment (youknowriad) [inline blocks/library/latest-posts/index.js]: I think we should use `import { Placeholder } from 'components';` to avoid code duplication when bundling the modules.

File: blocks/library/latest-posts/index.js [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+		constructor() {
+			super( ...arguments );
+
+			this.state = {
+				latestPosts: [],
+			};
+
+			const { poststoshow } = this.props.attributes;
+
+			getLatestPosts( poststoshow )
```
Comment (youknowriad) [inline blocks/library/latest-posts/index.js]: We should probably abort the query if the component is unmounted before getting the response (to avoid React warnings)

File: blocks/library/latest-posts/index.js [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+				.then( latestPosts => this.setState( { latestPosts } ) );
+		}
+
+		renderPostsLoading() {
```
Comment (youknowriad) [inline blocks/library/latest-posts/index.js]: Personal opinion, feel free to ignore. 

I prefer to avoid these render* functions when they are simple enough. It makes the component easier to refactor IMO (transform to functional or anything else when needed).

Comment (aduth) [inline blocks/library/latest-posts/index.js]: And in the case they're not simple, it's a good sign the component could do to be broken up into multiple smaller self-contained components.

File: blocks/library/latest-posts/index.js (hunk 8 above)
Comment (lamosty) [inline blocks/library/latest-posts/index.js]: I think yes. However, how can I import `wp.*`? `wp` is a global variable which is already present so I'm not sure how I would import it.

Comment (youknowriad) [inline blocks/library/latest-posts/index.js]: `import { __ } from 'i18n';` This would work, all the `wp.*` modules defined in Gutenberg are configured to allow importing them like this.

**Comment by lamosty** [comment]
In c9c7146, I added a basically blank full post content fixtures. since latest posts block returns `null` in `save`, it doesn't save anything in database. This is what I added in `core-latestposts.html` file:

```
<!-- wp:core/latestposts poststoshow="5" --><!-- /wp:core/latestposts -->
```

But that doesn't test anything (except the `poststoshow` attribute). Should we instead use the HTML which gets rendered from the server-side rendering?

File: blocks/library/latest-posts/index.js (hunk 10 above)
Comment (lamosty) [inline blocks/library/latest-posts/index.js]: Fixed in 64b30f1.

However, I'm not sure how I can easily test/trigger `componentWillUnmount`. Do you have some idea? :)

File: blocks/library/latest-posts/index.js (hunk 8 above)
Comment (lamosty) [inline blocks/library/latest-posts/index.js]: Ah, perfect, thanks!

File: blocks/library/latest-posts/index.js (hunk 10 above)
Comment (youknowriad) [inline blocks/library/latest-posts/index.js]: Is there any reason to store the request in the state. I tend to use simple class properties when changes to this property don't have to trigger a rerender.

> However, I'm not sure how I can easily test/trigger componentWillUnmount. Do you have some idea? :)

yep, hard to test. I'd suggest throttling the network requests and removing the block before the response (and ensure no warnings are shown in the console)

**Review by youknowriad — APPROVED** [review]
Aside avoiding to store the request in the state, this looks solid to me 👍

**Comment by jasmussen** [comment]
Nice work! 🎉

File: blocks/library/latest-posts/index.php [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+ *
+ * @param array $attributes The block attributes.
+ *
+ * @return string Returns the post content with latest posts added.
+ */
+function gutenberg_block_core_latest_posts( $attributes ) {
+	$posts_to_show = 5;
+
+	if ( array_key_exists( 'poststoshow', $attributes ) ) {
+		$posts_to_show = $attributes['poststoshow'];
```
Comment (westonruter) [inline blocks/library/latest-posts/index.php]: Since there is no sanitization or validation of block attributes (yet), it should probably be done here, ensuring it is a number and that it is greater than 0, but less than (maybe) 100. 

See also https://github.com/WordPress/gutenberg/issues/886#issuecomment-305518238

**Comment by lamosty** [comment]
Thank you very much for all the great tips and reviews! @youknowriad @notnownikki @aduth @mtias @westonruter @jasmussen 

I've learned quite a lot here and refreshed my JS skills. :)

**Comment by mtias** [comment]
Thank you @lamosty for working on this one.

File: gutenberg.php [hunk 13]
```diff
@@ -9,6 +9,8 @@
  * @package gutenberg
  */
 
+define( 'GUTENBERG__PLUGIN_DIR', plugin_dir_path( __FILE__ ) );
```
Comment (nylen) [inline gutenberg.php]: These `define`s are duplicating code already in place elsewhere:  https://github.com/WordPress/gutenberg/blob/3f727da24ee9059759d4ab4669a6681839219a49/lib/client-assets.php#L13-L22

The `GUTENBERG__` (double underscore) convention is unique to the constants introduced here; others in the codebase do not use this convention.  It seems unnecessary to me.

I don't have a preference on whether this is done as a function or constants (the function is pretty fast), but it needs to be consistent in approach and naming.

Also, `gutenberg.php` should contain the bare minimum necessary to initialize the plugin, because it is going to be rewritten during the build process.  These defines should probably happen in a new `lib/constants.php` or similar.

Comment (lamosty) [inline gutenberg.php]: Ah, I didn't notice that, my bad.

In a new PR, I'll fix those things.

Thanks!

File: lib/blocks.php [hunk 14]
```diff
@@ -128,3 +130,12 @@ function do_blocks( $content ) {
...
 add_filter( 'the_content', 'do_blocks', 10 ); // BEFORE do_shortcode().
+
+/**
+ * Loads the server-side rendering of blocks. If your block supports
+ * server-side rendering, add it here.
+ */
+function gutenberg_load_blocks_server_side_rendering() {
+	require_once GUTENBERG__BLOCKS_LIBRARY_DIR . '/latest-posts/index.php';
+}
+add_action( 'init', 'gutenberg_load_blocks_server_side_rendering' );
```
Comment (nylen) [inline lib/blocks.php]: I think there are a few other issues here, not least that `npm run package-plugin` is currently broken because this file was not added to the plugin zip package.  I'll open a PR to try to fix.
